packages/sdk/src/cross-chain-messenger.ts

```diff
@@ -1212,29 +1258,85 @@ export class CrossChainMessenger {
+          gameCount.sub(1),
+          Math.min(100, gameCount.toNumber())
         )
-    } catch (err) {
-      if (err.message.includes('L2OutputOracle: cannot get output')) {
+
+      // Find a game with a block number that is greater than or equal to the block number that the
+      // message was included in. We can use this proposal to prove the message to the portal.
+      const match = latestGames
+        .map((game: any) => {
+          const [blockNumber] = ethers.utils.defaultAbiCoder.decode(
+            ['uint256'],
+            game.extraData
+          )
+          return {
+            ...game,
+            l2BlockNumber: blockNumber,
+          }
+        })
+        .find((game: any) => {
+          return game.l2BlockNumber.gte(resolved.blockNumber)
+        })
+
+      // TODO: It would be more correct here to actually verify the proposal since proposals are
+      // not guaranteed to be correct. proveMessage will actually do this verification for us but
+      // there's a devex edge case where this message appears to give back a valid proposal that
+      // ends up reverting inside of proveMessage. At least this is safe for users but not ideal
+      // for developers and we should work out the simplest way to fix it. Main blocker is that
+      // verifying the proposal may require access to an archive node.
+
+      // If there's no match then we can't prove the message to the portal.
+      if (!match) {
         return null
-      } else {
-        throw err
       }
-    }
 
-    // Now pull the proposal out given the output index. Should always work as long as the above
-    // codepath completed successfully.
-    const proposal = await this.contracts.l1.L2OutputOracle.getL2Output(
-      l2OutputIndex
-    )
+      // Put the result into the same format as the old logic for now to reduce added code.
+      l2OutputIndex = match.index
+      proposal = {
+        outputRoot: match.rootClaim,
+        timestamp: match.timestamp,
+        l2BlockNumber: match.l2BlockNumber,
+      }
+    } else {
+      // Try to find the output index that corresponds to the block number attached to the message.
+      // We'll explicitly handle "cannot get output" errors as a null return value, but anything else
+      // needs to get thrown. Might need to revisit this in the future to be a little more robust
+      // when connected to RPCs that don't return nice error messages.
+      try {
+        l2OutputIndex =
+          await this.contracts.l1.L2OutputOracle.getL2OutputIndexAfter(
+            resolved.blockNumber
+          )
+      } catch (err) {
+        if (err.message.includes('L2OutputOracle: cannot get output')) {
+          return null
+        } else {
+          throw err
+        }
+      }
+
+      // Now pull the proposal out given the output index. Should always work as long as the above
+      // codepath completed successfully.
+      proposal = await this.contracts.l1.L2OutputOracle.getL2Output(
```

Comment:
Only case where this could possibly be undefined is if the output is deleted, in that case you'll just get a runtime error. Not ideal but whatever. Also this is for the legacy codepath that will no longer be relevant after FPAC ships.

Reply:
Sgtm